### ISSUE TITLE
reportUnboundVariable false error: Fails on check of argument used in inner function

### ISSUE THREAD

**Issue by tzmara** [microsoft/pyright#3535]
**Describe the bug**
pyright complains about unbound Variable. but it is bound and working. This only happens if I do have a statement in an inner function.

**To Reproduce**
- Write functin which has 1 argument (`arg`) and returns an inner function
- The inner function uses the argument (`arg`) from outer function in return statement
- (works until here)
- Has a statement with param (e.g. `if not arg: arg = 42`)

**Expected behavior**
This shouldn't be THIS kind of error.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
# THIS IS BROKEN
def broken(param=None):
    def inner():
        if not param:    # error: "param" is unbound (reportUnboundVariable)
            param = ""
        return param
    return inner

# THIS IS NOT BROKEN
def working(param=None):
    def inner():
        return param
    return inner
```

**VS Code extension or command-line**
cli: `pyright --version == 1.1.251`
python: `python --version == 3.7.13`

**Additional context**
It's also broken on deeper nesting of functions (with the `if` part in the inner functions) but not, when the `if` part is in the outer function.

**Comment by erictraut** [comment]
Pyright is correct here. You can see this if you add the following line to your example and run it in the Python interpreter:
```python
broken()()
```

This results in a runtime exception "UnboundLocalError: local variable 'param' referenced before assignment".

The reason is that `param` is a local variable in the `inner`. If your intent is for `param` to refer to the parameter of `broken`, you would need to include a `nonlocal param` statement in `inner`.

**Comment by tzmara** [comment]
Thank you! Learned something and sorry für mistagging and misinterpreting this :).

**Comment by erictraut** [comment]
No worries. The rules that Python uses for binding are quite counterintuitive for cases like this.
